tests/unit/GenericGraph.cpp

```diff
@@ -617,3 +617,234 @@ BOOST_AUTO_TEST_CASE(MutableEdgeNodeRemovalTest) {
   revng_check(C->successorCount() == 0);
   revng_check(C->predecessorCount() == 1);
 }
+
+// Tests for the UndirectedChildIterator implementation
+BOOST_AUTO_TEST_CASE(MutableEdgeNodeUndirectedIteratorTest) {
+  GenericGraph<MutableEdgeNode<std::string, size_t>> Graph;
+  using std::distance;
+  using std::next;
+
+  // We create a diamond shaped graph
+  auto *A = Graph.addNode("A");
+  auto *B = Graph.addNode("B");
+  auto *C = Graph.addNode("C");
+  auto *D = Graph.addNode("D");
+
+  A->addSuccessor(B, 1);
+  A->addSuccessor(C, 2);
+  B->addSuccessor(D, 3);
+  C->addSuccessor(D, 4);
+
+  auto ASuccessorEdges = A->successor_edges();
+  auto ASuccessorDistance = distance(ASuccessorEdges.begin(),
+                                     ASuccessorEdges.end());
+  auto APredecessorEdges = A->predecessor_edges();
+  auto APredecessorDistance = distance(APredecessorEdges.begin(),
+                                       APredecessorEdges.end());
+  auto AUndirectedEdges = A->undirected_child_edges();
+  auto AUndirectedDistance = distance(AUndirectedEdges.begin(),
+                                      AUndirectedEdges.end());
+  auto Count = A->successorCount() + A->predecessorCount();
+
+  // We test that each node as the expected successors, predecessors, and
+  // undirected edges
+  revng_check(2 == ASuccessorDistance);
+  revng_check(0 == APredecessorDistance);
+  revng_check(2 == AUndirectedDistance);
+  revng_check(next(AUndirectedEdges.begin(), Count) == AUndirectedEdges.end());
+
+  // We check the order in which successors are visited
+  auto ASuccessorIt = ASuccessorEdges.begin();
+  auto AEdgeView = *ASuccessorIt;
+  revng_check(AEdgeView.Neighbor == B);
+  revng_check(*AEdgeView.Label == 1);
+  ASuccessorIt++;
+  AEdgeView = *ASuccessorIt;
+  revng_check(AEdgeView.Neighbor == C);
+  revng_check(*AEdgeView.Label == 2);
+  ASuccessorIt++;
+  revng_check(ASuccessorIt == ASuccessorEdges.end());
+
+  // We check the order in which predecessors are visited
+  auto APredecessorIt = APredecessorEdges.begin();
+  revng_check(APredecessorIt == APredecessorEdges.end());
+
+  // We check that the undirected neighbors are the successors and predecessors
+  // in order
+  auto AUndirectedIt = AUndirectedEdges.begin();
+  AEdgeView = *AUndirectedIt;
+  revng_check(AEdgeView.Neighbor == B);
+  revng_check(*AEdgeView.Label == 1);
+  AUndirectedIt++;
+  AEdgeView = *AUndirectedIt;
+  revng_check(AEdgeView.Neighbor == C);
+  revng_check(*AEdgeView.Label == 2);
+  AUndirectedIt++;
+  revng_check(AUndirectedIt == AUndirectedEdges.end());
```

Comment:
You do the same tests for all the nodes A B C and D.

Please refactor this in a helper function that takes a single node as argument, and the constants you want to test (e.g. 1 and 2 for this case).

Reply:
I factored all this code in an helper https://github.com/revng/revng/blob/ff2b7e90cbf0ff106a7853aa468bfff9920cc855/tests/unit/GenericGraph.cpp#L621

and I also factored the code to check the DFS order visits https://github.com/revng/revng/blob/ff2b7e90cbf0ff106a7853aa468bfff9920cc855/tests/unit/GenericGraph.cpp#L684 https://github.com/revng/revng/blob/ff2b7e90cbf0ff106a7853aa468bfff9920cc855/tests/unit/GenericGraph.cpp#L695

Reply:
:+1: